torch/csrc/autograd/python_variable.cpp

```diff
@@ -509,17 +509,15 @@ void initTensorImplConversion(PyObject* module) {
     auto p = c10::intrusive_ptr<c10::TensorImpl, at::UndefinedTensorImpl>::
         unsafe_reclaim_from_nonowning(static_cast<c10::TensorImpl*>(ptr));
     AT_CHECK(p.defined(), "Can't wrap undefined tensor");
-    AT_CHECK(!p->is_variable(), "Can wrap only non-variable tensor");
```

Comment:
I found that we need to allow passing with-autograd-meta tensors into Caffe2, for the following reasons:

In Caffe2's `TestWorkspace::testTorchInterop`: https://github.com/pytorch/pytorch/blob/bc5398451e5de266767120a18f75e08f06d4bf8e/caffe2/python/workspace_test.py#L307-L311
The `z = torch.ones((4,), dtype=torch.int64)` line makes `z` a Variable (which has AutogradMeta), and the `workspace.FeedBlob('bar', z)` line passes `z` into Caffe2's workspace. If we were to preserve the previous invariant of passing only no-autograd-meta tensors into Caffe2, there are two probable approaches:

1. Remove `z`'s AutogradMeta in place, and pass it into the Caffe2 workspace.
Problem: the `z[0,1] = 123` line after `workspace.RunOperatorOnce()` in the above code block expects `z` to be a Variable (with AutogradMeta). If we remove `z`'s AutogradMeta in the earlier step, `z` will not be a Variable anymore, and this will cause segfault on the `z[0,1] = 123` line.

2. Shallow-copy `z`, and pass the copy into the Caffe2 workspace.
Problem: The `Reshape` operator (executed via `workspace.RunOperatorOnce(core.CreateOperator("Reshape", ['bar'], ['bar', '_'], shape=(2,2)))`) changes the sizes and strides of `bar`. The original semantics before this PR is that these changes will also be reflected in `z`. However, if `bar` is now a shallow-copy of `z`, changes to sizes and strides of `bar` will not be reflected in `z`, which makes the `z[0,1] = 123` line fail because `z` still has the old sizes `(4,)`, not the new sizes `(2, 2)` that we want it to have.

Due to the above reasons, we should allow passing with-autograd-meta tensors into Caffe2, and instead always use `at::AutoNonVariableTypeMode` thread-local guard to make sure gradients are not being accumulated.

Reply:
I don't think this test matters and it doesn't follow the python API anyway, right?  I.e. this is logically calling `.data` and `.data` doesn't propagate changes back to the original tensor anyway.

CC @dzhulgakov to check whether the test above actually matters.

Reply:
wait, can we do:
1) but, reattach the autograd meta when we get back to pytorch?  (by calling the autograd meta).